tools/match.py

```diff
@@ -0,0 +1,189 @@
+	# after all these `def`s and `class`es, the body continues
+	def advance():
+		nonlocal curr, curr_val
+		curr, curr_val = next(tokens, (END_OF_PATTERN, None))
+	def accept(accepted):
+		if curr == accepted:
+			advance()
+			return True
+		return False
+	def expect(expected):
+		if not accept(expected):
+			raise UserError('Unexpected %r' % curr)
+	def expect_lit():
+		if curr_val is None:
+			raise UserError('Unexpected %r' % curr)
+		result = curr_val
+		advance()
+		return result
+	# S -> '*' | DISJ
+	def parse_s():
+		if accept('*'):
+			return Conj() # empty conjunction always holds
+		else:
+			return parse_disj()
+	# DISJ -> CONJ (('|' | '+') CONJ)*
+	def parse_disj():
+		result = Disj([parse_conj()])
+		while accept('|') or accept('+'):
+			result.append(parse_conj())
+		return result
+	# CONJ -> LL ('&' LL)*
+	def parse_conj():
+		result = Conj([parse_ll()])
+		while accept('&'):
+			result.append(parse_ll())
+		return result
+	# LL -> '(' DISJ ')' | '!' LL | str | str '~' str
+	def parse_ll():
+		if accept('('):
+			result = parse_disj()
+			expect(')')
+			return result
+		if accept('!'):
+			return Not(parse_ll())
+		# otherwise:
+		s = expect_lit()
+		if accept('~'):
+			return StrRange(s, expect_lit())
+		else:
+			return SubString(s)
+	# disjunction: True iff at least one of its items is True
+	class Disj(list):
+		def apply(self, string):
+			return any(item.apply(string) for item in self) # recall that `self` is a list
+	# disjunction: True iff all its items are True
+	class Conj(list):
+		def apply(self, string):
+			return all(item.apply(string) for item in self) # just replaced `any` by `all`
+	# negation: True iff its subformula is False
+	class Not:
+		def __init__(self, formula):
+			self.formula = formula
+		def apply(self, string):
+			return not self.formula.apply(string)
+	# range predicate: True iff at least one prefix of given string is lexicographically between the two bounds
+	class StrRange:
+		def __init__(self, a, b):
+			if a > b:
+				raise UserError('Invalid range: %r comes after %r' % (a, b))
+			self.a = a
+			self.bn = b + '\x7f' # 'DEL' - highest ASCII char - we want "zzz" in "a~z"
+		def apply(self, string):
+			return self.a <= string <= self.bn
+	# substring predicate: (for some fixed s) True iff given string contains s as its substring
+	class SubString:
+		def __init__(self, substring):
+			self.substring = substring
+		def apply(self, string):
+			return self.substring in string
+	# here is the promised body
+	return parse_s()
+
+def load_pattern(string):
+	return parse_pattern(scan_pattern(string))
+
+####
+
+def main():
+	args = sys.argv[1:]
+	opts = None
+	
+	def scan_options(longopts_str):
+		nonlocal opts, args
+		opts_list, args_list = gnu_getopt(args, '', longopts_str.split())
+		opts_dict = dict(opts_list)
+		if len(opts_dict) != len(opts_list):
+			raise UserError('Duplicate option!')
+		opts = opts_dict
+		args = args_list
+	
+	scan_options('help as=')
+	
+	if '--help' in opts:
```

Comment:
I don't understand, what this tool does. Sorry. It's obvious it works with some expressions, but that's all I understood.